src/webnn_native/ops/Reduce.cpp

```diff
@@ -40,24 +40,57 @@ namespace webnn_native { namespace op {
         }
     }
 
-    MaybeError Reduce::Validate() {
-        MaybeError maybeError = OperatorBase::Validate();
+    MaybeError Reduce::CalculateShape() {
+        auto inputShape = mInputs[0]->Shape();
+        std::vector<int32_t> reducedShape(inputShape.size()), outputShape;
+        reducedShape = inputShape;
+        auto axes = mAxes;
+        for (size_t i = 0; i < axes.size(); ++i) {
+            if (axes[i] == -1) {
+                axes[i] = inputShape.size() - 1;
+            }
+            reducedShape[axes[i]] = 1;
+        }
+
+        if (!mOptions.keepDimensions) {
+            for (size_t i = 0; i < inputShape.size(); ++i) {
+                // Don't keep reduced dimensions with size of 1.
+                if (!(inputShape[i] != 1 && reducedShape[i] == 1)) {
```

Comment:
You can delete the 1 with `axes` from outputShape, 
like 
```
for (size_t i = axes.size() -1; i >0 ; --i) {
outputShape.earse(outputShape.begin() +  axes[i]);
}
```
Maybe you can merge the loop into L.48 and remove `reducedShape` local variable

Reply:
Thanks, but earse will change the vector iterator so we can't delete multiple elements. So I use push_back to keep the unreduced axes. @fujunwei

Reply:
Please earse it from end to begin as my code prsentation.

Reply:
> Please earse it from end to begin as my code prsentation.
I think the axes may be in unexpected order such as {0,3,1,2}. Then although from end to begin, we will erase wrong elements,right?

Reply:
Good point, let's keep it, thanks for explaination.

Reply:
> Good point, let's keep it, thanks for explaination.

Thanks, I may need add some annotations here to explain.